inference-engine/src/gna_plugin/transformations/insert_transpose_before_matmul.cpp

```diff
@@ -0,0 +1,64 @@
+// Copyright (C) 2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "transformations/insert_transpose_before_matmul.hpp"
+
+#include <ngraph/opsets/opset1.hpp>
+#include <ngraph/pattern/op/wrap_type.hpp>
+#include <ngraph/pattern/op/or.hpp>
+
+using namespace GNAPluginNS;
+
+NGRAPH_RTTI_DEFINITION(InsertTransposeBeforeMatmul, "InsertTransposeBeforeMatmul", 0);
+
+InsertTransposeBeforeMatmul::InsertTransposeBeforeMatmul() {
+    auto reshape = ngraph::pattern::wrap_type<ngraph::opset1::Reshape>();
+    auto matmul1 = ngraph::pattern::wrap_type<ngraph::opset1::MatMul>({ngraph::pattern::any_input(), reshape});
+    auto matmul2 = ngraph::pattern::wrap_type<ngraph::opset1::MatMul>({reshape, ngraph::pattern::any_input()});
+    auto root = std::make_shared<ngraph::pattern::op::Or>(ngraph::OutputVector{matmul1, matmul2});
+
+    ngraph::matcher_pass_callback callback = [=](ngraph::pattern::Matcher &m) {
+        auto& pattern_map = m.get_pattern_value_map();
+        auto reshape_node = pattern_map.at(reshape).get_node_shared_ptr();
+        auto reshape_in_shape = reshape_node->get_input_shape(0);
+        auto reshape_out_shape = reshape_node->get_output_shape(0);
+        if (reshape_in_shape.front() == reshape_out_shape.front()) {
+            return false;
+        }
+
+        if (reshape_out_shape.size() != 2 || reshape_out_shape[0] == 1 || reshape_out_shape[1] == 1) {
```

Comment:
I think it should be included in the pattern. For example, I suppose we can use `pattern::rank_equals` here

Reply:
I replaced the first check (reshape_out_shape.size() != 2 with pattern::rank_equals in the pattern. But I'm not sure that 2 other checks can be moved here.